### PR TITLE
Popover: show add filter button needing no inner scroll

### LINKED ISSUE #14307: I shouldn't have to scroll to see the add-filter button in the notebook's add-filter popover
When trying to filter on a field where you can either select values from a list or via a search box, the add-filter popover seems to have a max-height that causes the add-filter button to get pushed down, requiring you to have to scroll to add the filter, which is annoying.

Dropdown list example:

have-to-scroll-again

Search box example:

add-filter-have-to-scroll

### PR DESCRIPTION
**Pull request by gusaiani**
Fixes #14307 

## How to Test

If doing manual exploration, please look for use cases not covered in this description, as popovers are used pervasively in our application and not yet 100% covered by tests. 

### Simple Question Column Headers

1. From home, click on `Ask a question`
2. Click on `Simple question`
3. Choose `Sample Dataset`
4. Choose any of the tables, for example `People`
5. Click on the top of the columns
6. Click on `Filter by this column`
7. Try the various filters.

<details><summary>Screenshot</summary>

</details>

### Custom Question Filter

1. From home, click on `Ask a question`
2. Click on `Custom question`
3. Choose `Sample Dataset`
4. Choose any of the tables, for example `People`
5. Click on `Add filters to narrow your answers`
6. Try all types of fields: text, location, datetime and so on. 

<details><summary>Screenshot</summary>

</details>

### Filter in a Saved Dashboard

<details><summary>Screen recording</summary>
<video src="https://user-images.githubusercontent.com/380816/119669010-0d7c4680-be0e-11eb-8f61-811739f0a131.mov" />
</details>

### Adding Number Parameter to Filter in a Saved Dashboard

See screen recording above and follow video below.

<details><summary>Screen recording</summary>
  <video src="https://user-images.githubusercontent.com/380816/119669675-9c895e80-be0e-11eb-9272-3c7323c68eb6.mov" />
</details>

### CODE REVIEW

File: frontend/src/metabase/components/FieldValuesWidget.jsx [hunk 1]
```diff
@@ -442,6 +448,7 @@ export class FieldValuesWidget extends Component {
           width: this.props.expand ? this.props.maxWidth : null,
           minWidth: this.props.minWidth,
           maxWidth: this.props.maxWidth,
+          marginBottom: 60,
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/components/FieldValuesWidget.jsx:451]: I tried gradually decreasing the margin locally and in the end got completely rid of it. Looks much better, imo.

**Comment by nemanjaglumac** [comment]
As @flamber noted in the original issue, this was mostly visible on smaller screens. This is a screenshot for the arbitrary resolution 1472 x 900px. Note how we still have both outter and inner scroll. It feels as if we should also make the "header" (the top part) fixed, it that's possible at all given in how many different places this component is used?

**Comment by gusaiani** [comment]
> When one scrolls (outer scroll) all the way to the bottom, now search box is not visible. Afaic, that creates almost the same UX inconvenience as having to scroll to add a filter.

Yes, I've had time to think about these things as I fished around the code.

Making both the top and bottom fixed may, in some cases, leave no space for scrolling in the options. It may end up too small.

This Popover has the visual constraint of opening either:

1. above the button that opened it, then growing up to, but not past, the top of the window
2. below the button that opened it, then growing down to, but not past, the bottom of the window

This is very neat in many many cases. Makes the product magical and stuff.
But we may be demanding a bit too much of the visual constraint with all the use cases.

File: frontend/src/metabase/components/FieldValuesWidget.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/components/FieldValuesWidget.jsx]: Thanks Nemanja. 
This marginBottom is entangled with another instance of the Popover on another page.
On this.

**Comment by mazameli** [comment]
Couple things I've noticed so far while testing on Chrome 90:

Add Filter button hidden behind the "Include today" checkbox's footer:

image

I'm seeing some unnecessary scrollbars here now at far right and at bottom:

image

In this state there's an unnecessary scrollbar on the far right:

image

File: frontend/src/metabase/components/FieldValuesWidget.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (gusaiani) [reply frontend/src/metabase/components/FieldValuesWidget.jsx:451]: Should be fixed.

**Comment by gusaiani** [comment]
@mazameli @nemanjaglumac, thanks for your findings.

Those should be treated now.

**Comment by nemanjaglumac** [comment]
@gusaiani following your repro steps for the filtering based on the column I found this:
https://share.getcloudapp.com/bLukolA6

- absolutely positioned element (container with the button) is overlapping with the dropdown content - the hint for that was cut outer scroll
    - also, noticed that this element has `box-sizing` set to `border-box` explicitly. It's not necessary because we set that at the root level so this element (as all others) will inherit that
- `bottom` is set to 0 - it should be `-$elementlheight` (that is - there should be the `padding-bottom` set to `$elementlheight` in the parent element)
- just wondering - why is `width: calc(100% - 2px)`? Where do we get those 2px from?

**Comment by gusaiani** [comment]
@nemanjaglumac thanks for the checks.

I will start looking into the specific findings.

About the CSS details, yes, I will work to make them nicer. Generally it's been about dealing with inner scrolling, bottom rounded corners and not going over the dimensions of the Popover (and making corners square, or messing with the drop shadow).

**Comment by gusaiani** [comment]
@nemanjaglumac The thing I believe would be most important to do at this point is a look at the entire application for places where the changes may be broken breaking things.

Just to get a sense of how many are left (since the ones covered by e2e tests seem ok).

If we go one find, one fix, we won't get a sense of how many we have in total (if it's 30 cases for example, we may be better off talking about this).

As I don't know the corners of this application, to say the least, I was wondering if you or @flamber or @mazameli would be willing and able to do it?

**Comment by nemanjaglumac** [comment]
> As I don't know the corners of this application, to say the least, I was wondering if you or @flamber or @mazameli would be willing and able to do it?

I will definitely try to help, but @daltojohnso and @howonlee probably have the freshest and more extensive understanding of where this component is used. I might not even realize it's used in some parts of the ui.

**Comment by gusaiani** [comment]
> I will definitely try to help, but @daltojohnso and @howonlee probably have the freshest and more extensive understanding of where this component is used. I might not even realize it's used in some parts of the ui.

Thanks, I will ping them.

**Comment by gusaiani** [comment]
> This probably doesn't have anything to do with your changes. Popover is being cut, depending on the screen width. Let me know if I should file a new issue around this:
https://share.getcloudapp.com/GGuWx7wo

Yes, this screenshot is from `stats`, happens there as well.
At the very least the whole page scrolls past the table, allowing users to click through.
Needs improvement for sure.

image

**Comment by ariya** [comment]
This has some notable refactoring, beyond just a bug fix, and therefore it should go to master instead.

**Comment by gusaiani** [comment]
@ariya Yup, repointing it to master.

As we don't have enough coverage on all the Popovers, even the fix going to master and undergoing bugsquashing and more time in manual exploration may save us from regressions.

**Comment by gusaiani** [comment]
Closed in favor of https://github.com/metabase/metabase/pull/16175, which is pointed towards master.

**Pull request closed without merging**